packages/sources/refund.move

```diff
@@ -0,0 +1,514 @@
+            base_pool: pool::new(ctx),
+            booster_pool: pool::new(ctx),
+            accounting: accounting::new(),
+            phase: 1,
+            timeout_ts: none()
+        };
+
+        transfer::public_transfer(publisher, sender);
+        transfer::share_object(list);
+    }
+
+    // === Phase 1: Setup ===
+
+    /// Adds addresses and corresponding amounts to the refund pool. This endpoint
+    /// is reserved to Aldrin, i.e the owner of the `Publisher` object
+    /// 
+    /// #### Panics
+    /// 
+    /// - If the publisher is not valid.
+    /// - If the lengths of the addresses and amounts vectors do not match.
+    /// - If a given address already exists in the refund pool
+    /// - If there are duplicated addresses in the `adresses` vector
+    public entry fun add_addresses(
+        pub: &Publisher,
+        pool: &mut RefundPool,
+        addresses: vector<address>,
+        amounts: vector<u64>,
+    ) {
+        assert_address_addition_phase(pool);
+        assert_publisher(pub);
+        assert!(vector::length(&addresses) == vector::length(&amounts), EAddressesAmountsVecLenMismatch);
+        
+        let len = vector::length(&addresses);
+
+        while (len > 0) {
+            let amount = vector::pop_back(&mut amounts);
+            
+            let total_to_refund = total_to_refund_mut(&mut pool.accounting);
+            *total_to_refund = *total_to_refund + amount;
+
+            table::add(
+                &mut pool.unclaimed,
+                vector::pop_back(&mut addresses),
+                amount,
+            );
+
+            len = len - 1;
+        };
+
+        vector::destroy_empty(addresses);
+        vector::destroy_empty(amounts);
+    }
+
+    // === Phase 2: Funding ===
+
+    /// Permissionless endpoint for funding the fund pool with the given SUI coin.
+    public entry fun fund(
+        pool: &mut RefundPool,
+        coin: Coin<SUI>,
+        ctx: &mut TxContext,
+    ) {
+        assert_funding_phase(pool);
+
+        let amount = coin::value(&coin);
+        refund_table::insert_or_add(funders_mut(&mut pool.base_pool), sender(ctx), amount);
+
+        // If overfunded then return the excess
+        let total_to_refund = total_to_refund(&pool.accounting);
+        let total_raised = total_raised_mut(&mut pool.accounting);
+
+        let is_overfunded = *total_raised + amount > total_to_refund;
+        assert!(!is_overfunded, 0);
+
+        *total_raised = *total_raised + amount;
+        balance::join(funds_mut(&mut pool.base_pool), coin::into_balance(coin)); 
+    }
+
+    // === Phase 3: Claim Refund ===
+
+    /// Allows a claimer to claim their refund from the pool. This endpoint
+    /// should be called by the address registered in the `unclaimed` table.
+    /// Transfers the claimed amount back to the claimer.
+    /// 
+    /// #### Panics
+    /// 
+    /// - If the claimer's address is not in the list of unclaimed refunds.
+    public entry fun claim_refund(
```

Comment:
Do we wanna allow claiming refunds after the timeout phase has elapsed? 

If not, a check for that would solve the problem mentioned on my other comment on this file.

Reply:
Implemented